### PR TITLE
Jelares adds image analysis prototype

### PR DESCRIPTION
**Pull request by jelares**
I updated my demo to use the new Images Manager

### CODE REVIEW

File: backstory/pom.xml [hunk 1]
```diff
@@ -68,6 +68,13 @@
       <scope>test</scope>
     </dependency>
 
+    <dependency>
```
Comment (UdayKalra) [inline backstory/pom.xml:71]: Could you switch this to mockito-inline so there's just the one dependency?

File: backstory/src/demo/java/com/google/sps/servlets/AnalyzeImageServlet.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    // path to the image, rather than the path returned by imagesService which contains a host.
+    try {
+      URL url = new URL(imagesService.getServingUrl(options));
+      return url.getPath();
+    } catch (MalformedURLException e) {
+      return imagesService.getServingUrl(options);
+    }
+  }
+
+  /** Returns image byte data from BlobKey of image stored with Blobstore. */
```
Comment (UdayKalra) [inline backstory/src/demo/java/com/google/sps/servlets/AnalyzeImageServlet.java:130]: I recommend using ```@param``` and ```@return``` in function javadoc headers for consistency with perspective and story.

File: backstory/pom.xml (hunk 1 above)
Comment (annemariecabs) [reply backstory/pom.xml:71]: I actually updated this in the fix-pom branch, which is now a part of master. It allows for the inline dependency, but still mainly has core listed in the dependencies list. A simple merge with master should solve this problem.

File: backstory/src/demo/java/com/google/sps/servlets/AnalyzeImageServlet.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+   * {@inheritDoc}
+   *
+   * Expecting a post request from Blobstore containing the data fields from the image-upload
+   * form. After having gone through Blobstore, the request will include the image uploaded. The
+   * form in the HTML will connect to the Blobstore URL, which encodes the image and then
+   * redirects the request to this Url. In this doPost, the image is analyzed with Vision API,
+   * and the analysis along with the image URL will be sent to datastore for permanent storage.
+   */
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form.
```
Comment (annemariecabs) [inline backstory/src/demo/java/com/google/sps/servlets/AnalyzeImageServlet.java:61]: What does get the input from the form refer to? I can't seem to locate the code it's describing.

File: backstory/src/demo/java/com/google/sps/servlets/AnalyzeImageServlet.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    final byte[] bytes = getBlobBytes(request, "image-upload");
+
+    if (bytes == null || imageUrl == null) {
+      // Redirect back to the HTML page.
+      response.sendError(400, "Please upload a valid image.");
+
+    } else {
+      // Gets the full label information from the image byte array by calling Images Manager.
+      ImagesManager demoImagesManager = new VisionImagesManager();
+      List<byte[]> imagesAsByteArrays = new ArrayList<>();
```
Comment (annemariecabs) [inline backstory/src/demo/java/com/google/sps/servlets/AnalyzeImageServlet.java:75]: Is there a reason why you don't have a type for this ArrayList. Shouldn't it be `ArrayList<byte[]`>?

File: backstory/src/demo/java/com/google/sps/servlets/data/AnalyzedImage.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+public final class AnalyzedImage {
+  private final String imageUrl;
+  private final String labelsJsonArray;
+
+  public AnalyzedImage(String imageUrl, String labelsJsonArray) {
+    this.imageUrl = imageUrl;
+    this.labelsJsonArray = labelsJsonArray;
+  }
+
+  /** Get the ImageUrl */
```
Comment (annemariecabs) [inline backstory/src/demo/java/com/google/sps/servlets/data/AnalyzedImage.java:27]: Shouldn't these getters have an `@return` tag to conform with the style guide?

File: backstory/src/demo/java/com/google/sps/servlets/data/AnalyzedImage.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+// limitations under the License.
+
+package com.google.sps.servlets.data;
+
+/** Represents an image analyzed with the Vison API */
+public final class AnalyzedImage {
+  private final String imageUrl;
+  private final String labelsJsonArray;
+
+  public AnalyzedImage(String imageUrl, String labelsJsonArray) {
```
Comment (annemariecabs) [inline backstory/src/demo/java/com/google/sps/servlets/data/AnalyzedImage.java:22]: Constructors should also probably be documented, even if it's only minimal documentation.

File: backstory/src/demo/webapp/vision-upload-prototype/vision-demo.css [hunk 7]
```diff
@@ -0,0 +1,23 @@
+/* General */
```
Comment (annemariecabs) [inline backstory/src/demo/webapp/vision-upload-prototype/vision-demo.css:1]: Nice job keeping to the style guide with your CSS!

File: backstory/src/demo/java/com/google/sps/servlets/AnalyzeImageServlet.java [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+      imagesAsByteArrays.add(bytes);
```
Comment (ykjchen) [inline backstory/src/demo/java/com/google/sps/servlets/AnalyzeImageServlet.java:76]: Nit: it looks like one of these utilities could be used to create an immutable list in one line:
https://howtodoinjava.com/java/collections/create-list-with-single-element/

File: backstory/src/demo/java/com/google/sps/servlets/AnalyzeImageServlet.java [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+      List<AnnotatedImage> annotatedImages = demoImagesManager.createAnnotatedImagesFromImagesAsByteArrays(imagesAsByteArrays);
+      // There wil only be one image uploaded at a time for the demo
```
Comment (ykjchen) [inline backstory/src/demo/java/com/google/sps/servlets/AnalyzeImageServlet.java:78]: typo: wil -> will

File: backstory/src/demo/java/com/google/sps/servlets/AnalyzeImageServlet.java [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    ByteArrayOutputStream outputBytes = new ByteArrayOutputStream();
+
+    int fetchSize = BlobstoreService.MAX_BLOB_FETCH_SIZE;
+    long currentByteIndex = 0;
+    boolean continueReading = true;
+    while (continueReading) {
+      // end index is inclusive, so we have to subtract 1 to get fetchSize bytes
+      byte[] b =
```
Comment (ykjchen) [inline backstory/src/demo/java/com/google/sps/servlets/AnalyzeImageServlet.java:159]: Naming this variable using a full word would make code below easier to read.

**Pull request merged**